source/agora/consensus/validation/Block.d

```diff
@@ -139,25 +141,12 @@ public string isInvalidReason (const ref Block block, Height prev_height,
             return fail_reason;
     }
 
-    Transaction[] expected_txs;
-    foreach(tx; block.txs)
-    {
-        if (tx.type == TxType.Coinbase)
-            continue;
-
-        // Poor mans deep copy
-        Transaction new_tx;
-        new_tx.type = tx.type;
-        new_tx.inputs = tx.inputs.dup;
-        new_tx.outputs = tx.outputs.dup;
-        new_tx.payload = tx.payload;
-        expected_txs ~= new_tx;
-    }
-
-    appendCoinbaseTX(expected_txs, block.header.missing_validators);
-
-    if (!isPermutation(expected_txs, block.txs))
-        return "Block: Invalid Coinbase transaction";
+    auto expected_cb_txs = getCoinbaseTX(block.txs,
+        block.header.missing_validators);
+    auto incoming_cb_txs = block.txs.filter!(
+            tx => tx.type == TxType.Coinbase);
+    if (!isPermutation(expected_cb_txs, incoming_cb_txs))
```

Comment:
I'm still not sure about this design. There can only be one coinbase per block, no? Why are there multiple ones being compared here?

Reply:
Incoming block could have more then one. It should only have the single expected coinbase tx.

Reply:
If you ask, why I didnt just check for `length == 1` is to make the check somewhat generic.

Reply:
I see. Well I guess it's less lines of code, although the actual intention is hard to figure out from the call.

The way I would design it is to always force the coinbase tx to be the first one. That does create issues with sorting, but I would basically only sort on `txs[1..$]`.

Anyway your approach works too.

Reply:
My solution would probably create issues with the merkle tree. So yours is better.

Reply:
I generally avoid stuff that may create edge cases that I can't immediately recognize. The most generalized solution I can think of is what I go with most of the time.